python/shotgun_data/shotgun_data_retriever.py

```diff
@@ -0,0 +1,723 @@
+
+        If thumb_path already contains a path then this method does nothing and just returns the path 
+        without further checking/work.
+
+        :param thumb_path:  Path to an existing thumbnail or None.
+        :param url:         The url for the thumbnail which may or may not still be valid!
+        :param entity_type: Type of the entity to retrieve the thumbnail for
+        :param entity_id:   Id of the entity to retrieve the thumbnail for
+        :param field:       The field on the entity that holds the url for the thumbnail to retrieve
+        :param load_image:  If True then if the thumbnail is downloaded from Shotgun then the file will 
+                            be loaded into a QImage
+        :returns:           A dictionary containing the cached path for the specified url and a QImage
+                            if load_image is True and the thumbnail exists in the cache.
+        """
+        if thumb_path:
+            # no need to do anything as the thumbnail was previously
+            # found when we ran the check!
+            return {}
+
+        # download the actual thumbnail. Because of S3, the url
+        # may have expired - in that case fall back, get a fresh url
+        # from shotgun and try again
+        thumb_path = self._get_thumbnail_path(url, self._bundle)
+        self._bundle.ensure_folder_exists(os.path.dirname(thumb_path))
+
+        # there may be a case where another process has alrady downloaded the thumbnail for us, so 
+        # make sure that we aren't doing any extra work :)
+        if not os.path.exists(thumb_path):
+
+            # try to download based on the path we have
+            try:
+                sgtk.util.download_url(self._shotgun_connection, url, thumb_path)
+            except TankError, e:
+                # Note: Unfortunately, the download_url will re-cast 
+                # all exceptions into tankerrors.
+                # get a fresh url from shotgun and try again
+                sg_data = self._shotgun_connection.find_one(entity_type, [["id", "is", entity_id]], [field])
+
+                if sg_data is None or sg_data.get(field) is None:
+                    # no thumbnail! This is possible if the thumb has changed
+                    # while we were queueing it for download.
+                    # indicate the fact that the thumbnail no longer exists on the server
+                    # by setting the path to None
+                    thumb_path = None
+
+                else:
+                    # download from sg
+                    url = sg_data[field]
+                    sgtk.util.download_url(self._shotgun_connection, url, thumb_path)
+
+            if thumb_path:
+                # now we have a thumbnail on disk, either via the direct
+                # download, or via the url-fresh-then-download approach
+                # the file is downloaded with user-only permissions
```

Comment:
Would read better if it said something like "The file is download with user-only permissions, so modify the permissions of the file so it's writable by others."